### PR TITLE
[SPARK-36323][SQL] Support ANSI interval literals for TimeWindow

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR proposes to support ANSI interval literals for `TimeWindow`.

### Why are the changes needed?

Watermark also supports ANSI interval literals so it's great to support for `TimeWindow`.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala [hunk 1]
```diff
@@ -110,10 +115,26 @@ object TimeWindow {
...
+          throw new IllegalArgumentException(ymIntervalErrMsg)
+        } else {
+          parsed
+        }
+      }
+    } catch {
+      case NonFatal(e) =>
+        throw QueryCompilationErrors.cannotParseTimeDelayError(interval, e)
     }
     cal.days * MICROS_PER_DAY + cal.microseconds
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:139]: Not related to the PR but:
1. `*` and `+` can overflow. I think we should use exact arithmetic ops here.
2. one day in `CalendarInterval` can be 23, 24, or 25 hours, see https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java#L42

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala [hunk 2]
```diff
@@ -110,10 +115,26 @@ object TimeWindow {
    *         precision.
    */
   def getIntervalInMicroSeconds(interval: String): Long = {
-    val cal = IntervalUtils.stringToInterval(UTF8String.fromString(interval))
-    if (cal.months != 0) {
-      throw new IllegalArgumentException(
-        s"Intervals greater than a month is not supported ($interval).")
+    val ymIntervalErrMsg = s"Intervals greater than a month is not supported ($interval)."
+    val cal = try {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:119]: This is similar to https://github.com/apache/spark/blob/07fa38e2c1082c2b69b3bf9489cee4dfe4db2c26/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L744-L758 . Could you put the code to a common helper function, and re-use it here and in `Dataset`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TimeWindowSuite.scala [hunk 3]
```diff
@@ -21,10 +21,10 @@ import org.scalatest.PrivateMethodTester
 
 import org.apache.spark.SparkFunSuite
 import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.util.DateTimeConstants._
 import org.apache.spark.sql.types.{LongType, StructField, StructType, TimestampNTZType, TimestampType}
 
 class TimeWindowSuite extends SparkFunSuite with ExpressionEvalHelper with PrivateMethodTester {
-
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TimeWindowSuite.scala:27]: Let's avoid unnecessary changes. This can cause conflicts in down stream.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TimeWindowSuite.scala [hunk 4]
```diff
@@ -147,4 +147,39 @@ class TimeWindowSuite extends SparkFunSuite with ExpressionEvalHelper with Priva
     assert(timestampNTZWindow.dataType == StructType(
       Seq(StructField("start", TimestampNTZType), StructField("end", TimestampNTZType))))
   }
+
+  test("SPARK-36323: Support ANSI interval literals for TimeWindow") {
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TimeWindowSuite.scala:151]: Could you test when `spark.sql.legacy.interval.enabled` is `true`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala (hunk 2 above)
Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:119]: At first I tried to put a common helper function but they are slightly different so I didn't.
But it's O.K, to try to factor out the similar code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala (hunk 1 above)
Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:139]: O.K, let's fix this in this PR.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TimeWindowSuite.scala (hunk 3 above)
Comment (sarutak) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TimeWindowSuite.scala:27]: Oops. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala (hunk 1 above)
Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala]: For `1`, I've fixed in this PR.
For `2`, it's about intervals, not timestamps so we don't need to consider daylight saving here right?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala]: > For 2, it's about intervals, not timestamps so we don't need to consider daylight saving here right?

ok. Let's keep the assumption of 24 hours per day here. Not sure it will work fine during daylight saving time but it seems we don't have enough test coverage for that at the moment.

Maybe, open an JIRA to test the case?

Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala]: Hmm, I'll check it out and open a JIRA.

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master/3.2.
Thank you, @sarutak .
